chore: Improves the Select component UI/UX - iteration 4

michael-s-molina opened the pull request:
### SUMMARY
Improves the Select component UI/UX - iteration 4.
- Migrates the native filters Select components to the new Select
- Fixes some bugs of the Select component related to searching and selection
- Simplifies the Select API for the async non-paginated version

Touched components:
```
src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx
src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx
src/filters/components/common.ts
src/filters/components/GroupBy/GroupByFilterPlugin.tsx
src/filters/components/Select/SelectFilterPlugin.tsx
src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx
src/filters/components/TimeGrain/TimeGrainFilterPlugin.tsx
```
@rusackas @villebro 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
https://user-images.githubusercontent.com/70410625/124596960-40682000-de39-11eb-8ca5-0cb26af8080b.mp4

### TESTING INSTRUCTIONS
- Enter in a dashboard
- Open the filter bar
- Open the native filter modal
- Check the selects of the native filter modal
- After adding some filters check the selects in the filter bar

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

michael-s-molina committed:
chore: Improves the Select component UI/UX - iteration 4

michael-s-molina commented:
@jinghua-qa @junlincc Can you help me test the native filters with the new select?

michael-s-molina commented:
/testenv up

michael-s-molina commented:
/testenv up

michael-s-molina commented:
/testenv up

geido commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts:
```diff
@@ -99,3 +100,28 @@ export const doesColumnMatchFilterType = (filterType: string, column: Column) =>
+  charts: ChartsState,
+) => {
+  const map = new Map<string, number>();
+  let mostUsedDatasouce = '';
```
Small typo here :)

geido commented:
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true

geido reviewed (commented):
The code is looking good! Just looking for the CI to pass and for the _env with native filters ON_ for some manual testing

rusackas commented:
Hmmm... something seems amiss, if I'm not mistaken. If I go to the Sales Dashboard, and create a new filter, the default dataset should be "cleaned_sales_data" but the Select now shows "27"
![image](https://user-images.githubusercontent.com/812905/125025036-29931a80-e03f-11eb-9f1f-7c2f5f88baa8.png)

"cleaned_sales_data" doesn't show up in the options list, but I _can_ search for it. When I select it, and create a filter, like so...
![image](https://user-images.githubusercontent.com/812905/125025125-5ba47c80-e03f-11eb-8d57-0850e9d17ef3.png)

I get this error in the filter bar:
![image](https://user-images.githubusercontent.com/812905/125025158-6828d500-e03f-11eb-8fdc-66f19589e476.png)

rusackas commented:
Well hmm... on my local build, with the Sales Dashboard, I see the correct data source in the Select. However, I can't create functional native filters (attempted value and numeric range) on that dashboard. On other dashboards, things are working seemingly fine. Something is amiss, but I'm not yet sure if it's the data source, the dashboard, or the feature(s).

michael-s-molina commented:
@rusackas Thanks for testing this. I'll investigate the problem with the Sales Dashboard.

michael-s-molina commented:
@rusackas 

> Hmmm... something seems amiss, if I'm not mistaken. If I go to the Sales Dashboard, and create a new filter, the default dataset should be "cleaned_sales_data" but the Select now shows "27"

Fixed

> I get this error in the filter bar:
> ![image](https://user-images.githubusercontent.com/812905/125025158-6828d500-e03f-11eb-8fdc-66f19589e476.png)

When I try to add a default value I get "Error: Time column "OrderDate" does not exist in dataset". So probably this dataset is corrupted and you can ignore this error.

michael-s-molina commented:
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true

rusackas commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:
```diff
@@ -45,23 +46,25 @@ import React, {
 import { Tooltip } from 'src/components/Tooltip';
+import { Radio } from 'src/components/Radio';
 import BasicErrorAlert from 'src/components/ErrorMessage/BasicErrorAlert';
+import rison from 'rison';
```
Teensy nit, but this could go above the `src/**/*` imports, with the other npm pkgs.

michael-s-molina replied:
Done.

rusackas commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/utils.ts:
```diff
@@ -99,3 +100,28 @@ export const doesColumnMatchFilterType = (filterType: string, column: Column) =>
+export const mostUsedDataset = (
+  datasets: DatasourcesState,
+  charts: ChartsState,
+) => {
```
Your logic here is all good... but for sport, this got me thinking... If you want to get fancy, you might be able to one-liner this with lodash magic... something like:
```
const mostUsedDataset = _(charts)
  .countBy('formData.datasource')
  .toPairs()
  .orderBy([1], ['desc'])
  .value().shift()[0];

return datasets[mostUsedDataset]?.id;
 ```
... but that's probably even more difficult to read 😜

michael-s-molina replied:
Really fancy! Cool functions from Lodash! I think readability is more important though. I had to open Lodash docs to understand how it worked 🤣 .

michael-s-molina commented:
@rusackas I want to do a small change before merging. I'll request your review again when it's done.

michael-s-molina commented:
@rusackas You can continue your review 😉

michael-s-molina commented:
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true

rusackas commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/FiltersConfigForm.tsx:
```diff
@@ -705,27 +706,25 @@ const FiltersConfigForm = (
+                  datasetDetails={datasetDetails}
+                  datasetId={initialDatasetId}
+                  onChange={(value: number) => {
+                    // We need reset column when dataset changed
```
```suggestion
                    // We need to reset the column when the dataset has changed
```

rusackas commented on superset-frontend/src/filters/components/GroupBy/GroupByFilterPlugin.tsx:
```diff
@@ -18,13 +18,11 @@
 import { FormItemProps } from 'antd/lib/form';
-import { Styles, StyledSelect, StyledFormItem, StatusMessage } from '../common';
+import { Select } from 'src/components';
+import { Styles, StyledFormItem, StatusMessage } from '../common';
```
Hmm... this "Styles" could probably use a more specific name.

michael-s-molina replied:
Done

rusackas commented on superset-frontend/src/filters/components/TimeColumn/TimeColumnFilterPlugin.tsx:
```diff
@@ -24,13 +24,11 @@ import {
+import { Select } from 'src/components';
 import { FormItemProps } from 'antd/lib/form';
-import { Styles, StyledSelect, StyledFormItem, StatusMessage } from '../common';
+import { Styles, StyledFormItem, StatusMessage } from '../common';
```
If you agreed with the first one, this `Styles` could also use a more specific/explicit name as well. But then again, these are nits, and I'd love to see this merged!

michael-s-molina replied:
Done

rusackas reviewed (approved):
LGTM, but it'd be nice to have more eyes on this, as it's chunky!

michael-s-molina commented:
@rusackas I'll do the suggested changes in a follow-up PR. @geido Can we get your approval as well?

geido commented:
Sure @michael-s-molina . I'll do some manual testing as well today and come back to you.

geido commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/ColumnSelect.tsx:
```diff
@@ -128,6 +128,7 @@ export function ColumnSelect({
     <Select
       mode={mode}
       value={mode === 'multiple' ? value || [] : value}
+      ariaLabel="Column select"
```
I think we could use the localization here

michael-s-molina replied:
Done

geido commented on superset-frontend/src/dashboard/components/nativeFilters/FiltersConfigModal/FiltersConfigForm/DatasetSelect.tsx:
```diff
@@ -0,0 +1,125 @@
+
+  return (
+    <Select
+      ariaLabel="Dataset"
```
I think localizing the `aria-label `would be good. I see other places as well where it is not being localized. I won't mention them all. Those are easy to spot.

michael-s-molina replied:
Done

geido reviewed (approved):
The latest changes to the code look good. Just a nit below. Now going for manual testing

geido reviewed (changes requested):
Hey @michael-s-molina please find the results of my manual testing efforts below:

I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.

<img width="1123" alt="Screen Shot 2021-07-19 at 15 25 47" src="https://user-images.githubusercontent.com/60598000/126167097-c5cf70be-b1c0-45e0-b9df-5165edc76b4d.png">

I don't see my newly created option "test" in the dropdown in multiple mode

<img width="931" alt="Screen Shot 2021-07-19 at 15 35 09" src="https://user-images.githubusercontent.com/60598000/126168220-6a8f2a6a-ae60-4afe-9af1-f04ee07bb018.png">

I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list

<img width="920" alt="Screen Shot 2021-07-19 at 15 41 50" src="https://user-images.githubusercontent.com/60598000/126169363-ea55db1e-cb09-4a16-b10f-8b805a57c53f.png">


I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.

<img width="1272" alt="Screen Shot 2021-07-19 at 15 39 38" src="https://user-images.githubusercontent.com/60598000/126168868-10c72d63-74d2-4479-bd64-aba8cece077e.png">

After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)

<img width="1058" alt="Screen Shot 2021-07-19 at 15 29 27" src="https://user-images.githubusercontent.com/60598000/126167491-c893a91e-1582-46aa-b6ea-96d21ee1b8ba.png">

michael-s-molina commented:
> Hey @michael-s-molina please find the results of my manual testing efforts below:
> 
> I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.
> 
> <img alt="Screen Shot 2021-07-19 at 15 25 47" width="1123" src="https://user-images.githubusercontent.com/60598000/126167097-c5cf70be-b1c0-45e0-b9df-5165edc76b4d.png">
> 
> I don't see my newly created option "test" in the dropdown in multiple mode
> 
> <img alt="Screen Shot 2021-07-19 at 15 35 09" width="931" src="https://user-images.githubusercontent.com/60598000/126168220-6a8f2a6a-ae60-4afe-9af1-f04ee07bb018.png">
> 
> I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list
> 
> <img alt="Screen Shot 2021-07-19 at 15 41 50" width="920" src="https://user-images.githubusercontent.com/60598000/126169363-ea55db1e-cb09-4a16-b10f-8b805a57c53f.png">
> 
> I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.
> 
> <img alt="Screen Shot 2021-07-19 at 15 39 38" width="1272" src="https://user-images.githubusercontent.com/60598000/126168868-10c72d63-74d2-4479-bd64-aba8cece077e.png">
> 
> After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)
> 
> <img alt="Screen Shot 2021-07-19 at 15 29 27" width="1058" src="https://user-images.githubusercontent.com/60598000/126167491-c893a91e-1582-46aa-b6ea-96d21ee1b8ba.png">

Thanks for the throughout testing. None of the issues are related to this PR. The first 3 are related to the Select component and will be fixed in separate PRs. The last 2 are isolated issues. Will open separate PRs also. So the plan is to open these separate PRs and then rebase this PR on top of the fixes. So let's leave this open for now.

michael-s-molina commented:
@geido @rusackas @junlincc 

> I have noticed a problem with the creation of new options. It is now allowing for the creation of existing options, which would then make unselecting any of the duplicated options impossible. It gets stuck.
> 
> I don't see my newly created option "test" in the dropdown in multiple mode
> 
> I don't see my newly created option in the dropdown in single mode. I think this used to work consistently with the single mode ON and allowNewOptions OFF, where the selected option is highlighted at top of the options list
> 

After investigation, I discovered that these were not bugs in the new `Select` component but problems in the `SelectFilterPlugin` component.  `SelectFilterPlugin` contained duplicated logic already present in the `Select` component. I removed the duplicated logic.

> I am not sure if this is related to this PR but I am unable to create hierarchical filters. The child filter never appears anywhere. Only the parent is showing up. When I go back to editing, the hierarchical option is always unchecked as if it wasn't saved.
> 

This was in fact related to this PR. Fixed.

> After adding a new filter, the previously selected options of the existing filters are not selected anymore. This may be because I haven't clicked "Apply", however, as a user I would expect that my filters remain filled as I add or delete other filters, even before applying (I am fine if this needs to be discussed outside of the scope of this PR)
> 

This one is by design, until now we only keep the values if the user clicks on "Apply". We can change this behavior if the product team agrees but if that's the case we need to open another low priority issue.

@rusackas @geido I also applied all your suggestions.

@geido Thanks for testing the features. I updated the environment and requested your review again.

michael-s-molina commented:
/testenv up FEATURE_DASHBOARD_NATIVE_FILTERS=true

geido commented:
Hey @michael-s-molina. Thanks for the fixes. There's only one problem that I could find. See below

When I choose _Time Grain_ and a default value, it won't allow me to select the option _Original value_. It fails to select it. This happens on both the modal and the left sidebar

<img width="1086" alt="Screen Shot 2021-07-20 at 17 12 03" src="https://user-images.githubusercontent.com/60598000/126349431-d693af54-2eb5-4884-829d-119b592a617c.png">

<img width="1102" alt="Screen Shot 2021-07-20 at 17 24 03" src="https://user-images.githubusercontent.com/60598000/126351138-580b12aa-45d0-4f3e-bc7e-bec89ae3d834.png">

michael-s-molina commented:
@geido This problem is already happening on master so I'll open another PR to fix it so we can merge this one.
